core/distributed/matrix.cpp

```diff
@@ -0,0 +1,429 @@
+    this->set_size(global_dim);
+
+    // temporary storage for the output
+    array<local_index_type> diag_row_idxs{exec};
+    array<local_index_type> diag_col_idxs{exec};
+    array<value_type> diag_values{exec};
+    array<local_index_type> offdiag_row_idxs{exec};
+    array<local_index_type> offdiag_col_idxs{exec};
+    array<value_type> offdiag_values{exec};
+    array<local_index_type> recv_gather_idxs{exec};
+    array<comm_index_type> recv_offsets_array{exec, num_parts + 1};
+
+    // build diagonal, off-diagonal matrix and communication structures
+    exec->run(matrix::make_build_diag_offdiag(
+        data, make_temporary_clone(exec, row_partition).get(),
+        make_temporary_clone(exec, col_partition).get(), local_part,
+        diag_row_idxs, diag_col_idxs, diag_values, offdiag_row_idxs,
+        offdiag_col_idxs, offdiag_values, recv_gather_idxs,
+        recv_offsets_array.get_data(), local_to_global_ghost_));
+
+    // read the local matrix data
+    const auto num_diag_rows =
+        static_cast<size_type>(row_partition->get_part_size(local_part));
+    const auto num_diag_cols =
+        static_cast<size_type>(col_partition->get_part_size(local_part));
+    const auto num_ghost_cols = local_to_global_ghost_.get_num_elems();
+    device_matrix_data<value_type, local_index_type> diag_data{
+        exec, dim<2>{num_diag_rows, num_diag_cols}, std::move(diag_row_idxs),
+        std::move(diag_col_idxs), std::move(diag_values)};
+    device_matrix_data<value_type, local_index_type> offdiag_data{
+        exec, dim<2>{num_diag_rows, num_ghost_cols},
+        std::move(offdiag_row_idxs), std::move(offdiag_col_idxs),
+        std::move(offdiag_values)};
+    as<ReadableFromMatrixData<ValueType, LocalIndexType>>(this->diag_mtx_)
+        ->read(diag_data);
+    as<ReadableFromMatrixData<ValueType, LocalIndexType>>(this->offdiag_mtx_)
+        ->read(offdiag_data);
+
+    // exchange step 1: determine recv_sizes, send_sizes, send_offsets
+    exec->get_master()->copy_from(exec.get(), num_parts + 1,
+                                  recv_offsets_array.get_data(),
+                                  recv_offsets_.data());
+    std::adjacent_difference(recv_offsets_.begin() + 1, recv_offsets_.end(),
+                             recv_sizes_.begin());
+    comm.all_to_all(recv_sizes_.data(), 1, send_sizes_.data(), 1);
+    std::partial_sum(send_sizes_.begin(), send_sizes_.end(),
+                     send_offsets_.begin() + 1);
+    send_offsets_[0] = 0;
+
+    // exchange step 2: exchange gather_idxs from receivers to senders
+    auto needs_host_buffer =
+        exec->get_master() != exec && !gko::mpi::is_gpu_aware();
+    if (needs_host_buffer) {
+        recv_gather_idxs.set_executor(exec->get_master());
+        gather_idxs_.clear();
+        gather_idxs_.set_executor(exec->get_master());
+    }
+    gather_idxs_.resize_and_reset(send_offsets_.back());
+    comm.all_to_all_v(recv_gather_idxs.get_const_data(), recv_sizes_.data(),
```

Comment:
maybe let all_to_all_v to handle the gpu_aware?

Reply:
This would require that we pass in an executor to the call, or the constructor of the communicator. At the moment, I wouldn't go that way, especially since that would complicate handling non-blocking calls quite a lot. Perhaps if we add the device guard to the mpi calls, then it makes sense to revisit this.